common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1217,6 +1264,20 @@ public boolean toByte(IntWrapper intWrapper) {
 
   @Override
   public String toString() {
+    byte[] bytes = getBytes();
+    // Optimization for ASCII characters: use deprecated string API which
+    // skips charset encoder and simply casts each byte into a char.
+    if (isAscii(bytes)) {
+      return new String(bytes, 0);
```

Comment:
I get that we might want to go off-road here and call a deprecated method, but does it help meaningfully? the benchmark doesn't have the resolution to determine it. Is it as fast to, maybe, pass US_ASCII as the charset in this case?

Reply:
That's a good question. I think it might be a little bit slower because the `US_ASCII` encoder needs to perform error-checking to validate whether the bytes are actually ASCII characters. That validation is probably cheap (just checking the sign / highest order bit of the byte) but it's still technically a bit more work.

Good point about benchmark resolution, especially for short strings: I probably need to run more benchmark iterations for short strings to smooth out variation / noise.

Reply:
it does have to check all bytes though? isn't this grow linearly to the length of bytes?